design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,276 @@
+with the signed certificate.
+
+One of the checks is called
+[`SecretOwnerReferenceValueMismatch`](https://github.com/cert-manager/cert-manager/blob/b78af1ef867f8776715cae3dd6a8b83049c4d9b2/internal/controller/certificates/policies/checks.go#L511)
+and checks that the owner reference on the Secret resource matches the one
+on the Certificate resource.
+
+### Test Plan
+
+<!---
+Describe how the new functionality will be tested (unit tests, integration tests (if applicable), e2e tests)
+-->
+
+### Graduation Criteria
+
+We propose to release this feature in GA immediately and skip the "beta"
+phase that consists of gathering user feedback, since this feature has a
+low user-facing surface. We think that we will be able to take a good
+decision (e.g., the name of the new field, whether it is a boolean or a
+string, and which values the field can take) while developing the feature
+in the PR.
+
+We don't think this feature needs to be [feature gated][feature gate].
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+
+### Upgrade / Downgrade Strategy
+
+Upgrading from a version without this feature to a version with this
+feature won't be breaking.
+
+Downgrading, however, will be a breaking change, since a new field will be
+introduced.
+
+### Supported Versions
+
+This feature will be supported in all the versions of Kubernetes that are
+supported by cert-manager.
+
+## Production Readiness
+
+<!--
+This section should confirm that the feature can be safely operated in production environment and can be disabled or rolled back in case it is found to increase failures.
+-->
+
+### How can this feature be enabled / disabled for an existing cert-manager installation?
+
+<!--
+
+Can the feature be disabled after having been enabled?
+
+Consider whether any additional steps will need to be taken to start/stop using this feature, i.e change existing resources that have had new field added for the feature before disabling it.
+
+
+Do the test cases cover both the feature being enabled and it being disabled (where relevant)?
+
+-->
+
+### Does this feature depend on any specific services running in the cluster?
+
+No.
+
+### Will enabling / using this feature result in new API calls (i.e to Kubernetes apiserver or external services)?
+
+No.
+
+### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+No.
+
+### Will enabling / using this feature result in significant increase of resource usage? (CPU, RAM...)
+
+No.
+
+## Drawbacks
+
+## Alternatives
+
+It is possible to use the
+[`csi-driver`](https://github.com/cert-manager/csi-driver) to circumvent
+the problem of "too many ephemeral Secret resources stored in etcd". Using
+the CSI driver, no Secret resource is created, alliviating the issue.
+
+Flant relies on Certificate resources to provide X.509 certificates for
+their customers. The certificates cannot be tied to a Pod workload.
```

Comment:
@RomanenkoDenys Did I understand correctly the reason why you can't use [`cert-manager/csi-driver`](https://github.com/cert-manager/csi-driver)?

Reply:
Hi! We can't use cert-manager/csi-driver because our customers have tons of different deploys with certificates. Our main goal is to do things that don't affect customers.

Reply:
Usage of [cert-manager/csi-driver](https://github.com/cert-manager/csi-driver) will be good solution in future, we will discuss it in. our team.

Reply:
I have rephrased the paragraph to say that Certificate is part of the "user-facing API" that Flant offers to customers, meaning that entirely moving to csi-driver isn't possible. Does it make sense?

Reply:
It's not a part of our API, but users knows how to use certificate resources and uses them. And they do not know anything about usage of csi-driver, so that solution complicates user's experience.